index.js

```diff
@@ -1,148 +1,264 @@
+      throw new Error(`${name} is not defined in the cache`)
+    }
+
+    // TODO validate key?
 
-function currentSecond () {
-  if (_currentSecond !== undefined) {
-    return _currentSecond
+    return this[kValues][name].get(key)
   }
-  _currentSecond = Math.floor(Date.now() / 1000)
-  setTimeout(_clearSecond, 1000).unref()
-  return _currentSecond
-}
 
-function _clearSecond () {
-  _currentSecond = undefined
+  async set (name, key, value, ttl, references) {
+    if (!this[kValues][name]) {
+      throw new Error(`${name} is not defined in the cache`)
+    }
+
+    // TODO validate key, value, ttl, references?
+
+    return this[kValues][name].set(key, value, ttl, references)
+  }
+
+  async invalidate (name, references) {
+    if (!this[kValues][name]) {
+      throw new Error(`${name} is not defined in the cache`)
+    }
+
+    // TODO validate references?
+
+    return this[kValues][name].invalidate(references)
+  }
 }
 
 class Wrapper {
-  constructor (func, key, serialize, cacheSize, ttl, onHit) {
-    this.ids = new LRUCache(cacheSize)
-    this.error = null
-    this.started = false
+  // TODO signature
+  constructor (func, name, serialize, references, storage, ttl, onDedupe, onHit, onMiss) {
+    this.dedupes = new Map()
     this.func = func
-    this.key = key
+    this.name = name
     this.serialize = serialize
+    this.references = references
+
+    this.storage = storage
     this.ttl = ttl
+    this.onDedupe = onDedupe
     this.onHit = onHit
-  }
-
-  buildPromise (query, args, key) {
-    query.promise = this.func(args, key)
-    // we fork the promise chain on purpose
-    const p = query.promise.catch(() => this.ids.set(key, undefined))
-    if (this.ttl > 0) {
-      query.cachedOn = currentSecond()
-    } else {
-      // clear the cache if there is no TTL
-      p.then(() => this.ids.set(key, undefined))
-    }
+    this.onMiss = onMiss
   }
 
   getKey (args) {
     const id = this.serialize ? this.serialize(args) : args
     return typeof id === 'string' ? id : stringify(id)
   }
 
+  getStorageKey (key) {
+    return `${this.name}~${key}`
+  }
+
+  getStorageName () {
+    return `${this.name}~`
+  }
+
   add (args) {
     const key = this.getKey(args)
-    const onHit = this.onHit
 
-    let query = this.ids.get(key)
+    let query = this.dedupes.get(key)
     if (!query) {
       query = new Query()
       this.buildPromise(query, args, key)
-      this.ids.set(key, query)
-    } else if (this.ttl > 0) {
-      onHit()
-      if (currentSecond() - query.cachedOn > this.ttl) {
-        // restart
-        this.buildPromise(query, args, key)
-      }
+      this.dedupes.set(key, query)
     } else {
-      onHit()
+      this.onDedupe(key)
     }
 
     return query.promise
   }
 
-  clear (value) {
+  /**
+   * wrap the original func to sync storage
+   */
+  async wrapFunction (args, key) {
+    const storageKey = this.getStorageKey(key)
+    const data = await this.storage.get(storageKey)
```

Comment:
Avoid the await if the returned value is not a promise.

Reply:
the `get` method in storage interface is an async function to cover redis io, so this is needed

Reply:
You can do the following:

```js
let data = this.storage.get(storageKey)
if (data && tyepof data.then === 'function') {
  data = await data
} 
```